File: Day_7_No_Space_Left_On_Device/Python/main.py
from __future__ import annotations


class Directory:
    def __init__(self, name: str, parent_directory: Directory = None):
        # name of directory
        self.name = name
        # parent directory object
        self.parent_directory = parent_directory
        # list of immediate subdirectory objects
        self.subdirectories = []
        # dictonary of file names to file sizes
        self.files = {}

    def add_subdirectory(self, subdirectory: Directory):
        """
        Add Directory as a subdirectory.
        """
        if id(self) == id(subdirectory):
            raise ValueError('cannot add self as subdirectory')

        self.subdirectories.append(subdirectory)

    def add_file(self, filename: str, filesize: int):
        """
        Add file with name and size.
        """
        self.files[filename] = filesize

    def get_subdirectory(self, sub_dirname: str) -> Directory:
        """
        Find subdirectory by name.
        """
        # raise exception if at least one matching subdirectory is not found
        try:
            subdirectory = [d for d in self.subdirectories if d.name == sub_dirname][0]
        except IndexError:
            raise ValueError(f'subdirectory with name {sub_dirname} not found')

        return subdirectory

    def get_parent_directory(self) -> Directory:
        """
        Get parent Directory.
        """
        return self.parent_directory

    def compute_size(self: Directory) -> int:
        """
        Recursively compute size of Directory's contents.
        """
        total_size = 0

        # recursively compute sizes of subdirectories
        for subdirectory in self.subdirectories:
            total_size += subdirectory.compute_size()

        # compute sizes of files
        for filesize in self.files.values():
            total_size += filesize

        return total_size


if __name__ == '__main__':
    # file with filesystem data
    file_path = '../filesystem.txt'
    # open and read file
    with open(file_path) as f:
        file_contents = f.read()

    TOTAL_DISK_SPACE = 70000000
    REQUIRED_DISK_SPACE = 30000000
    DIRECTORY_SIZE_THRESHOLD = 100000

    # create root directory
    root = Directory(name='/')
    # list of directories navigated
    navigated_directories = [root]
    # current directory
    pwd = root
    # represents whether or not to expect ls command output
    ls_mode = False

    for line in file_contents.splitlines():
        line_split = line.split()

        # if line is a command (i.e. cd, ls)
        if line_split[0] == '$':
            cmd = line_split[1]

            if cmd == 'cd':
                ls_mode = False
                dirname = line_split[2]

                # navigate to root directory
                if dirname == '/':
                    pwd = root
                # navigate to parent directory
                elif dirname == '..':
                    pwd = pwd.get_parent_directory()
                # navigate to subdirectory
                else:
                    pwd = pwd.get_subdirectory(dirname)
            elif cmd == 'ls':
                # turn on ls mode output
                ls_mode = True
                continue
        # if line is not a command and in ls mode
        elif ls_mode:
            # if directory listed
            if line_split[0] == 'dir':
                dirname = line_split[1]
                # create new directory with current directory as parent
                new_dir = Directory(name=dirname, parent_directory=pwd)
                # add new directory as subdirectory of current directory
                pwd.add_subdirectory(new_dir)
                # add new directory to list of navigated directories
                navigated_directories.append(new_dir)
            # else if file listed
            else:
                filename = line_split[1]
                filesize = int(line_split[0])
                # add file name and size to directory
                pwd.add_file(filename, filesize)
        # raise exception if unknown line encountered
        else:
            raise ValueError(f'encountered non-command input {line} while ls_mode = False')

    # compute sizes of navigated directories
    navigated_directory_sizes = [d.compute_size() for d in navigated_directories]
    # get sum of directory sizes under threshold
    print(sum([size for size in navigated_directory_sizes if size <= DIRECTORY_SIZE_THRESHOLD]))

    # compute size of root directory
    total_size = root.compute_size()
    # compute unused space
    unused_space = TOTAL_DISK_SPACE - total_size
    # compute amount of space needed
    extra_space_needed = REQUIRED_DISK_SPACE - unused_space
    # get size of smallest directory that can be removed to get space needed
    print(min([size for size in navigated_directory_sizes if size >= extra_space_needed]))

File: Day_7_No_Space_Left_On_Device/Go/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

const TotalDiskSpace = 70000000
const RequiredDiskSpace = 30000000
const DirectorySizeThreshold = 100000

var rootDirectory *Directory

// struct representing each file
type File struct {
	// file name
	Name string
	// file size
	Size int
}

// struct representing each directory
type Directory struct {
	// directory name
	Name string
	// parent directory pointer
	Parent *Directory
	// hash of subdirectory name to subdirectory pointer
	Subdirectories map[string]*Directory
	// hash of file name to file pointer
	Files map[string]*File
	// cached value of directory size
	// used to avoid computing size multiple times
	// -1 represents empty value
	CachedSize int
}

// directory constructor
func Mkdir() *Directory {
	// create new directory
	directory := &Directory{}
	// initialize hash of subdirectories
	directory.Subdirectories = make(map[string]*Directory)
	// initialize hash of files
	directory.Files = make(map[string]*File)
	// set cached size to empty value
	directory.CachedSize = -1

	return directory
}

// add file to directory
func (directory *Directory) Touch(file *File) error {
	// check and return error if file already exists
	_, ok := directory.Files[file.Name]
	if ok {
		return fmt.Errorf("file with name %s already exists in directory %s", file.Name, directory.Name)
	}

	// store file pointer in hash
	directory.Files[file.Name] = file

	return nil
}

// add subdirectory to directory
func (directory *Directory) Mv(subdirectory *Directory) error {
	// check and return error if file already exists
	_, ok := directory.Subdirectories[subdirectory.Name]
	if ok {
		return fmt.Errorf("subdirectory with name %s already exists in directory %s", subdirectory.Name, directory.Name)
	}

	// store subdirectory pointer in hash
	directory.Subdirectories[subdirectory.Name] = subdirectory
	// set directory as parent of subdirectory
	subdirectory.Parent = directory

	return nil
}

// navigate to directory
func (directory *Directory) Cd(directoryName string) (*Directory, error) {
	// navigate to root directory if absolute root path
	if directoryName == "/" {
		return rootDirectory, nil
	}

	// navigate to parent directory if relative parent path
	if directoryName == ".." {
		if directory.Parent == nil {
			return nil, fmt.Errorf("directory %s has no parent", directory.Name)
		}

		return directory.Parent, nil
	}

	// navigate to subdirectory
	subdirectory, ok := directory.Subdirectories[directoryName]
	if !ok {
		return nil, fmt.Errorf("no subdirectory with name %s exists in directory %s", directoryName, directory.Name)
	}

	return subdirectory, nil
}

// recursively compute size of directory files and subdirectories
func (directory *Directory) Du() int {
	// if directory cached size is non-empty, return cached size
	if directory.CachedSize >= 0 {
		return directory.CachedSize
	}

	size := 0

	// add file sizes
	for _, file := range directory.Files {
		size += file.Size
	}

	// recursively add subdirectory sizes
	for _, subdirectory := range directory.Subdirectories {
		size += subdirectory.Du()
	}

	// set directory cached size
	directory.CachedSize = size

	return size
}

// get line-by-line file scanner
func GetFileScanner(filePath string) (*bufio.Scanner, error) {
	f, err := os.Open(filePath)
	if err != nil {
		return nil, fmt.Errorf("failed to open file %s, %s", filePath, err.Error())
	}

	scanner := bufio.NewScanner(f)

	return scanner, nil
}

func main() {
	// create root directory
	rootDirectory = Mkdir()
	rootDirectory.Name = "/"
	// set root directory to current directory
	pwd := rootDirectory
	// list of directories navigated
	directories := make([]*Directory, 0)
	// represents whether or not to expect ls command output in input
	lsMode := false
	// file with filesystem data
	filePath := "../filesystem.txt"

	scanner, err := GetFileScanner(filePath)
	if err != nil {
		fmt.Println("failed to get file scanner", err.Error())
		return
	}

	// scan file line-by-line
	for scanner.Scan() {
		input := scanner.Text()
		inputSplit := strings.Fields(input)

		// if input is a command (i.e. cd, ls)
		if inputSplit[0] == "$" {
			cmd := inputSplit[1]

			if cmd == "cd" {
				lsMode = false
				// navigate to directory
				directoryName := inputSplit[2]
				pwd, err = pwd.Cd(directoryName)
			} else if cmd == "ls" {
				// turn on ls mode output
				lsMode = true
			} else {
				panic(fmt.Sprintf("encountered unknown command input %s", input))
			}
		} else if lsMode {
			// if directory listed
			if inputSplit[0] == "dir" {
				directoryName := inputSplit[1]

				// create new directory
				newDirectory := Mkdir()
				newDirectory.Name = directoryName

				// add new directory as subdirectory of current directory
				err := pwd.Mv(newDirectory)
				if err != nil {
					panic(fmt.Sprintf("failed to convert add directory %s, %s", directoryName, err.Error()))
				}

				// add new directory to navigated directories
				directories = append(directories, newDirectory)
			} else {
				fileName := inputSplit[1]

				// convert file size from string to int
				fileSize, err := strconv.Atoi(inputSplit[0])
				if err != nil {
					panic(fmt.Sprintf("failed to convert %s into integer, %s", inputSplit[0], err.Error()))
				}

				// create new file
				newFile := &File{
					Name: fileName,
					Size: fileSize,
				}

				// add new file to current directory
				err = pwd.Touch(newFile)
				if err != nil {
					panic(fmt.Sprintf("failed to convert add directory %s, %s", fileName, err.Error()))
				}
			}
		} else {
			panic(fmt.Sprintf("encountered non-command input %s while lsMode = False", input))
		}
	}

	// compute sizes of navigated directories
	directorySizes := make([]int, len(directories))
	for i := range directories {
		directorySizes[i] = directories[i].Du()
	}

	// get size of directories smaller than threshold
	totalSizeUnderThreshold := 0
	for _, size := range directorySizes {
		if size <= DirectorySizeThreshold {
			totalSizeUnderThreshold += size
		}
	}

	fmt.Println(totalSizeUnderThreshold)

	// compute size of root directory
	rootSize := rootDirectory.Du()
	// compute unused space
	unsusedSpace := TotalDiskSpace - rootSize
	// compute amount of space needed
	extraSpaceNeeded := RequiredDiskSpace - unsusedSpace

	// compute size smallest directory that can be removed to get space needed
	smallestRemovableDirectorySize := rootSize
	for _, size := range directorySizes {
		if size >= extraSpaceNeeded && size < smallestRemovableDirectorySize {
			smallestRemovableDirectorySize = size
		}
	}

	fmt.Println(smallestRemovableDirectorySize)
}
